### PR TITLE
[IMPROVED] NumPending calculations and subject index memory in filestore

### PR DESCRIPTION
**Pull request by derekcollison**
Swapped out psim as a go hashmap for our stree implementation.

Stree is an adaptive radix tree implementation used for storing and retrieving literal subjects. It also allows quick matching to wildcard subjects, which is it's major design goal along with using less memory in high subject cardinality situations.

This will be used in the filestore implementation to replace the PSIM hash map which was fast at insert and lookup but suffered when trying to filter based on wildcard subjects.

This is used specifically in calculations on NumPending with a wildcard, and given we push folks to use larger muxed streams with down filtered consumers and/or mirrors this was becoming a performance issue.

Signed-off-by: Derek Collison <derek@nats.io>

### CODE REVIEW

File: server/stree/node16.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// Deletes a child from the node.
+func (n *node16) deleteChild(c byte) {
+	for i, last := uint16(0), n.size-1; i < n.size; i++ {
+		if n.keys[i] == c {
+			// Unsorted so just swap in last one here, else nil if last.
+			if i < last {
+				n.keys[i] = n.keys[last]
+				n.children[i] = n.children[last]
```
Comment (neilalexander) [inline server/stree/node16.go:79]: Should it be cleaning up `n.keys[last]` and `n.children[last]` here?

File: server/stree/node16.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// Shrink if needed and return new node, otherwise return nil.
+func (n *node16) shrink() node {
+	if n.size > 4 {
+		return nil
+	}
+	nn := newNode4(nil)
+	for i := uint16(0); i < n.size; i++ {
+		nn.addChild(n.keys[i], n.children[i])
```
Comment (neilalexander) [inline server/stree/node16.go:97]: `node256` has a check here for if `n.children[i] != nil`, might make sense to do the same here too.

File: server/stree/node16.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+// Will match parts against our prefix.no
+func (n *node16) matchParts(parts [][]byte) ([][]byte, bool) {
+	return matchParts(parts, n.prefix[:n.prefixLen])
+}
+
+// Iterate over all children calling func f.
+func (n *node16) iter(f func(node) bool) {
+	for i := uint16(0); i < n.size; i++ {
+		if !f(n.children[i]) {
```
Comment (neilalexander) [inline server/stree/node16.go:110]: Similarly `node256` has a check here for `n.children[i] != nil` too before the iterator call.

File: server/stree/node4.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// Deletes a child from the node.
+func (n *node4) deleteChild(c byte) {
+	for i, last := uint16(0), n.size-1; i < n.size; i++ {
+		if n.keys[i] == c {
+			// Unsorted so just swap in last one here, else nil if last.
+			if i < last {
+				n.keys[i] = n.keys[last]
+				n.children[i] = n.children[last]
```
Comment (neilalexander) [inline server/stree/node4.go:78]: Same here with cleaning up `n.keys[last]` and `n.children[last]` as above.

File: server/stree/node4.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+// Will match parts against our prefix.
+func (n *node4) matchParts(parts [][]byte) ([][]byte, bool) {
+	return matchParts(parts, n.prefix[:n.prefixLen])
+}
+
+// Iterate over all children calling func f.
+func (n *node4) iter(f func(node) bool) {
+	for i := uint16(0); i < n.size; i++ {
+		if !f(n.children[i]) {
```
Comment (neilalexander) [inline server/stree/node4.go:105]: And the `nil` check here too.

**Review by neilalexander — COMMENTED** [review]
Overall looks good! Few minor things I noticed.

File: server/stree/node16.go (hunk 1 above)
Comment (derekcollison) [reply server/stree/node16.go:79]: Its gated by n.size, so no need, but we could. Was trying to optimize for speed. Will make the change though.

File: server/stree/node16.go (hunk 2 above)
Comment (derekcollison) [reply server/stree/node16.go:97]: node256 can have holes, node4 and node16 are gated by n.size which in node4 and node16 control that loop.

File: server/stree/node16.go (hunk 3 above)
Comment (derekcollison) [reply server/stree/node16.go:110]: Again that is because those node types are gated by n.size and are all positioned at the beginning. Node256 has fixed slots and we need to check for nil there. They can also be spread out, so n.size does not give us those guarantees. Notice node256 does not have keys, its implicit into the children slice.

File: server/stree/node4.go (hunk 4 above)
Comment (derekcollison) [reply server/stree/node4.go:78]: Note comment from above but will add as well.

File: server/stree/node4.go (hunk 5 above)
Comment (derekcollison) [reply server/stree/node4.go:105]: Again not needed based on the way node4 and node16 work.

File: server/stree/node16.go (hunk 2 above)
Comment (neilalexander) [reply server/stree/node16.go:99]: Ah OK, got it!

**Review by neilalexander — APPROVED** [review]
LGTM

File: server/stree/dump.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+package stree
+
+import (
+	"fmt"
+	"io"
+	"strings"
+)
+
+// TODO(dlc) - This I used alot for debugging and testing. Probably should be made to return a string vs using fmt.Printf.
```
Comment (Jarema) [inline server/stree/dump.go:22]: This comment seems irrelevant

**Review by Jarema — APPROVED** [review]
Still doing deeper dive into some parts of the codebase and benching it, but code itself looks good!

LGTM!

File: server/stree/dump.go (hunk 6 above)
Comment (derekcollison) [reply server/stree/dump.go:22]: You are correct, the original version did not pass Travis so I converted it to take a io.Writer, will remove. Thanks.

File: server/stree/bench_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+func BenchmarkSubjectTreeMatch(b *testing.B) {
+	st := NewSubjectTree[struct{}]()
+
+	for i := 0; i < b.N; i++ {
+		// Do 100 objects for 2nd token.
+		subj := fmt.Sprintf("subj.%d.%d", rand.Intn(100)+1, i)
+		st.Insert([]byte(subj), struct{}{})
+	}
+
+	b.Logf("Running with %d subjects", b.N*b.N)
```
Comment (neilalexander) [inline server/stree/bench_test.go:32]: Minor & purely cosmetic but `b.N*b.N` is no longer correct here now that the second loop was taken out.

**Review by neilalexander — APPROVED** [review]
Otherwise LGTM!

File: server/stree/bench_test.go (hunk 7 above)
Comment (derekcollison) [reply server/stree/bench_test.go:32]: Actually though I removed that whole file, since I did the other one. Will remove.

**Review by wallyqs — COMMENTED** [review]
LGTM
